app/components/contacts/AddNewContact.js

```diff
@@ -0,0 +1,222 @@
+  static getDerivedStateFromProps(props, prevState) {
+    if (props.publicWalletAddress && !prevState.publicWalletAddress) {
+      return { publicWalletAddress: props.publicWalletAddress };
+    }
+    return null;
+  }
+
+  render() {
+    const { modalMode } = this.props;
+    const { publicWalletAddress, nickname, addressErrorMsg, nicknameErrorMsg, emailErrorMsg, renderKey } = this.state;
+    const isSaveButtonDisabled = !publicWalletAddress || !nickname || !!addressErrorMsg || !!nicknameErrorMsg || !!emailErrorMsg;
+    return (
+      <Wrapper showBorder={!modalMode} key={`render_key_${renderKey}`}>
+        {!modalMode && (
+          <TitleWrapper>
+            <Title>Add New Contact</Title>
+          </TitleWrapper>
+        )}
+        {this.renderFields()}
+        <ButtonWrapper alignment={modalMode ? 'right' : 'left'}>
+          <SmButton text="Save to contacts" isDisabled={isSaveButtonDisabled} theme="orange" onPress={this.handleSave} style={{ width: 164 }} />
+        </ButtonWrapper>
+      </Wrapper>
+    );
+  }
+
+  renderFields = () => {
+    return <React.Fragment>{fields.map((field: Field) => this.renderSingleField(field))}</React.Fragment>;
+  };
+
+  renderSingleField = ({ title, placeholder, fieldName, errorFieldName }: Field) => {
+    const { publicWalletAddress } = this.props;
+    const { addressErrorMsg, nicknameErrorMsg, emailErrorMsg } = this.state;
+    const isPublicAddressFromProps = fieldName === 'publicWalletAddress' && !!publicWalletAddress;
+    const errorMessages = {
+      publicWalletAddress: addressErrorMsg,
+      nickname: nicknameErrorMsg,
+      email: emailErrorMsg
+    };
+    return (
+      <FieldWrapper key={fieldName}>
+        <TextWrapper>
+          <BoldText>{title}</BoldText>
+        </TextWrapper>
+        <SmInput
+          type="text"
+          isDisabled={isPublicAddressFromProps}
+          placeholder={placeholder}
+          defaultValue={isPublicAddressFromProps ? publicWalletAddress : ''}
+          errorMsg={errorMessages[fieldName]}
+          onChange={({ value }) => this.handleTyping({ value, fieldName, errorFieldName })}
+        />
+      </FieldWrapper>
+    );
+  };
+
+  getErrorMessage = ({ fieldName, value }: { fieldName: string, value: string }) => {
+    const errorMessages = {
+      publicWalletAddress: 'Address is invalid',
+      nickname: 'Nickname is required',
+      email: 'Must enter a valid email'
+    };
+    const validators = {
+      publicWalletAddress: this.validateAddress,
+      nickname: this.validateNickname,
+      email: this.validateEmail
+    };
+    return validators[fieldName](value) ? null : errorMessages[fieldName];
+  };
+
+  handleTyping = ({ value, fieldName, errorFieldName }: { value: string, fieldName: string, errorFieldName: string }) => {
+    this.setState({ [fieldName]: value, [errorFieldName]: this.getErrorMessage({ fieldName, value }) });
+  };
+
+  handleSave = () => {
+    const { addToContacts, resolve } = this.props;
```

Comment:
why you use function named resolve? There are no promises resolved by calling this function so why using that name?

Reply:
Resolve in the literal sense of the word. To decide on. Resolving a value. 
This is a bad name because of promises...
I'll change to onSave.